File: AdrianBarros/quiz/q1.py
'''
Problem Statement
Given an array of positive numbers and a positive number k, find the maximum sum of any contiguous subarray of size k.
'''

'''
input: unsorted array + int
output: max sum
step1: start declaring two pointers
step2: loop through the array adding all the k numbers
step3: compare with the last max number and keep going
setp4: substract the first pointer to it and move both pointers
step5: return the max value


time complexity O(n+n) = O(n) two Separate loops but we only go over the whole loop once
space complexity: O(1)

edge cases: 
not as much as an edge case as a follow up question is what if we add neg numbers in the array:
        then we change max from 0 to -inf
what is if the array is empty or the number is 0 or neg
            return 0
'''


def max_sum_sub(arr, num):
    if not arr or num < 1:
        return 0
    ptr1 = ptr2 = 0
    max_sum = 0
    curr_sum = 0
    for i in range(num):
        curr_sum += arr[i]
        ptr2 += 1

    max_sum = curr_sum

    while ptr2 < len(arr):
        curr_sum -= arr[ptr1]
        ptr1 += 1
        curr_sum += arr[ptr2]
        max_sum = max(max_sum, curr_sum)
        ptr2 += 1
    return max_sum


print(max_sum_sub([1, 2, 3, 4, 5, 6], 3))  # 15
print(max_sum_sub([100, 90, 18, 4, 5, 6], 2))  # 190

File: alanEappen/quiz/q1.js
//Given an array of positive numbers and a positive number ‘k,’ find the maximum sum of any contiguous subarray of size ‘k’.

/*
Given: an array of positive numbers [1,2,3,4] and K =2
Goal: find the maximum subarray of size k


first create a function maxSub() that takes in the 
positive numbers and k

we then compute sum of the first window
let sum = 0
a for loop to add the contents into sum variable
for(let i = 0 i<k i++)
{
    sum+= num[i]
}

we then compute the next windows by removing
the first element of the last windowm and then adding it 

currSum = sum get the currentsum
currSum += num[i] - arr[i-k] removing the first element and adding to currentsum
then we compare both sums and return the max

edge cases{
    if arrray is empty return -1
    if array length is less than k return -1
}

*/


function maxSub(k, arr) {
    let Sum = 0;

    if (arr.length < k || arr.length === null) {
        return -1
    }

    for (let i = 0; i < k; i++) {//First window
        Sum += arr[i];
    }

    let currSum = Sum
    for (let i = k; i < arr.length; i++) {//Second Window
        currSum += arr[i] - arr[i - k];
        Sum = Math.max(Sum, currSum);
    }

    return Sum
}
console.log(
    `Maximum sum of a subarray of size K: ${maxSub(
        4,
        [1, 5, 6, 3, 7, 2] //21
    )}`
);

console.log(
    `Maximum sum of a subarray of size K: ${maxSub(
        4,
        [] //-1
    )}`
);

console.log(
    `Maximum sum of a subarray of size K: ${maxSub(
        4,
        [1, 3, 4] //-1
    )}`
);
